### PR TITLE
Mixed data model compute with proxy accessor

### PR DESCRIPTION
**Pull request by kxl-adsk**
This PR enables `mixed data model compute` between DG and USD, i.e. it allows DG connections to read and write to USD attributes. `ProxyAccessor` it the class responsible for providing this capability, which becomes the foundation to workflows like:

- parenting (e.g. DAG to Prim)
- animation (i.e. driving prim attributes with Maya animation curves)
- attribute level constraints

We decided to put `ProxyAccessor` capabilities on `ProxyShape` but the class itself can be used with any `MPxNode` with `ProxyStageProvider` interface for the querying stage and time. Centralizing all writes and reads from the USD stage allows deterministic evaluation order and avoids race conditions during parallel execution. 

The `ProxyAccessor` supports CachedPlayback, i.e. accelerated playback, proper cache invalidation when manipulating and keying accessor plugs and background compute.

This PR doesn't enable native workflows with UFE. A set of reference python methods is provided to illustrate how a particular workflow could be implemented.

The best place to start the review of this PR is the [ProxyAccessor](https://github.com/Autodesk/maya-usd/blob/kxl-adsk/MAYA-105180/mixed_data_model_compute/lib/mayaUsd/nodes/proxyAccessor.h#L49) class.

The best place to learn more about usage and how to use new capabilities provided by `ProxyShape` is [regression test file](https://github.com/Autodesk/maya-usd/blob/kxl-adsk/MAYA-105180/mixed_data_model_compute/test/lib/testMayaUsdProxyAccessor.py). You will see that we leverage proxyAccessor.py which is there only for testing & illustration purpose.

Currently known limitations
- Proxy shapes with in-memory root layer are not supported by Cached Playback background computation (see skipped testOutputForInMemoryRootLayer_Cache test)
- Data translation for a specific attribute type may not be available. ProxyAccessor is relying on `Converter` (#432) class, refer to its documentation to learn more about currently supported types.
- Modifying USD directly without Maya's commands doesn't allow animation (see skipped testKeyframeWithUFE_NoCaching and testKeyframeWithUSD_NoCaching test)
- Cyclic dependencies between accessor outputs and inputs
- Nice name for accessor attributes is not currently set for channels of compound attributes. We rely on `UsdMayaReadUtil::FindOrCreateMayaAttr` and this is where the fix will have to be done.
- We still have some room to improve the performance by caching more things in acceleration structure hold by `ProxyAccessor`

**ProxyAccessor is only enabled in PreviewReleases (starting with version 115) and currently only adsk plugin registers it.**

### CODE REVIEW

File: test/lib/testMayaUsdProxyAccessor.py [hunk 1]
```diff
@@ -0,0 +1,1265 @@
...
+from mayaUsd import lib as mayaUsdLib
+from mayaUsd.lib import GetPrim
+from mayaUsd.lib import proxyAccessor as pa
+
+from pxr import Usd, UsdGeom, Sdf, Tf, Vt
+
+from maya.debug.emModeManager import emModeManager
+from maya.plugin.evaluator.CacheEvaluatorManager import CacheEvaluatorManager, CACHE_STANDARD_MODE_EVAL
+
+class NonCachingScope(object):
```
Comment (kxl-adsk) [inline test/lib/testMayaUsdProxyAccessor.py:35]: All these helper classes and methods will have to be refactored. It's not part of this PR and we have another PR that moves tests out of Pixar plugin into our global test folder. We will coordinate on where test utils for caching and UFE should live after this PR.

**Comment by cfmoore007** [comment]
@kxl-adsk - I just tried this out and having some issues here.

Unit test fails, as per :  [testsFailed.txt](https://github.com/Autodesk/maya-usd/files/4804610/testsFailed.txt)

I think because of this, nothing seems to be working in Maya :

1) no feedback from :
```
from pxr import Tf
Tf.Debug.SetDebugSymbolsByName("USDMAYA_PROXYACCESSOR", 1)
```
2) then, attempts at the following did not work either :
- can't take a ufe object and move it under the DAG locator I have
- can't use the constrain command with both selected
- can't key position of ufe prims

Testing environment :
- Windows 10
- PR 115
- maya-usd dev @f45f050
- usd dev @e7ff8e8

**Comment by kxl-adsk** [comment]
As per comments in the description, there is no UFE workflow part currently available. Testing with regular parent or key commands are expected to not work. This PR delivers the compute side only. 

For failing test, I saw it on our PF (only Windows). 

In your log it looks like the failing test is testParentingDagObjectUnderUsdPrim_Caching. Non cached playback version passed...for some reason cache didn’t fill in background. 

Please filter out this test locally (use @unittest.skip(“”)) for the time being.

**Comment by cfmoore007** [comment]
Running parts of the .py script manually seems to work.

MixedDataModelTest1

MixedDataModelTest2

The proxyShape is driving the rotate and translation of the USD Prims
The locator follows the USD Prims as well.
MixedDataModelTest3

**Comment by cfmoore007** [comment]
> testParentingDagObjectUnderUsdPrim_Caching

100% tests passed, 0 tests failed out of 27

Label Time Summary:
ufe                  = 100.28 sec*proc (19 tests)
usdPreviewSurface    =  10.02 sec*proc (2 tests)

Total Test time (real) = 135.85 sec

Skipping that test worked.

**Comment by cfmoore007** [comment]
It appears in order to use the GetOrCreateAccessPlug() method on the proxyAccessor, the 'ufeItemParent' that is passed in must have a timeSample.

This example works ( modified from the [testMayaUsdProxyAccessor.py](https://raw.githubusercontent.com/Autodesk/maya-usd/kxl-adsk/MAYA-105180/mixed_data_model_compute/test/lib/testMayaUsdProxyAccessor.py) script )

[MixedDataModel_works.txt](https://github.com/Autodesk/maya-usd/files/4806282/MixedDataModel_works.txt)


"
// Confirm and test
ufeItemParent
// Result: <ufe.PyUfe.SceneItem object at 0x000001FF641457D8> # 
translatePlugParent   = pa.GetOrCreateAccessPlug(ufeItemParent, usdAttrName='xformOp:translate')
"

However, the following does not ( it's basically the same code, except no timeSampled valuse are added to the ufeItemParent )

[MixedDataModel_doesnot_work.txt](https://github.com/Autodesk/maya-usd/files/4806285/MixedDataModel_doesnot_work.txt)


"
// Test and confirm
ufeItemParent
// Result: <ufe.PyUfe.SceneItem object at 0x0000015D15DCF1F0> # 
translatePlugParent   = pa.GetOrCreateAccessPlug(ufeItemParent, usdAttrName='xformOp:translate')

E Error: ArgumentError: file F:\APPLICATIONS\MayaUSD\f45f050-mixed_data_model_compute\lib\python\mayaUsd\lib\proxyAccessor.py line 99: Python argument types in
    ReadUtil.FindOrCreateMayaAttr(NoneType, Variability, str, str, str)
did not match C++ signature:
    FindOrCreateMayaAttr(class pxrInternal_v0_20__pxrReserved__::SdfValueTypeName typeName, enum pxrInternal_v0_20__pxrReserved__::SdfVariability variability, class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > nodeName, class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > attrName, class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > attrNiceName='')  
"

**Comment by kxl-adsk** [comment]
You won’t be able to create access plugs for attributes that doesn’t exist on a prim. This is what happens in your “doesnot_work” example.
Important to mention that data type for access plug will be taken from usd attribute. 

Without providing attribute, you can fetch world matrix. Here is one example of it:

'''
worldMatrixPlugSphere = pa.GetOrCreateAccessPlug(ufeItemSphere, '', Sdf.ValueTypeNames.Matrix4d)
'''

File: test/lib/testMayaUsdProxyAccessor.py [hunk 2]
```diff
@@ -0,0 +1,1265 @@
...
+        unloaded = cmds.unloadPlugin(rs)
+        return (unloaded[0] == rs)
+
+    return True
+
+def makeUfePath(dagPath, sdfPath=None):
+    """
+    Make ufe item out of dag path and sdfpath
+    """
+    ufePath = ufe.PathString.path('|world{},{}'.format(dagPath,sdfPath) if sdfPath != None else '|world{}'.format(dagPath))
```
Comment (ppt-adsk) [inline test/lib/testMayaUsdProxyAccessor.py:231]: "|world" prefix is optional when creating a path using Ufe::PathString::path(), and should preferably be omitted.

File: test/lib/testMayaUsdProxyAccessor.py [hunk 3]
```diff
@@ -0,0 +1,1265 @@
...
+        # Validate that input accessor plugs wrote the data to USD
+        # Validate that output accessor plug has data driven by input accessor plug
+        
+        cachingScope.checkValidFrames(self, self.cache_empty)
+        cachingScope.waitForCache(self)
+        cachingScope.checkValidFrames(self, self.cache_allFrames)
+        
+        cmds.currentTime(1)
+        v1 = cmds.getAttr('{}.{}'.format(nodeDagPath,worldMatrixPlugSphere))
+        v2 = ufeTransform3dCube.segmentInclusiveMatrix()
```
Comment (ppt-adsk) [inline test/lib/testMayaUsdProxyAccessor.py:502]: We are using segmentInclusiveMatrix() to get the matrix of the final segment of the UFE path, correct?  I.e. in this case, the USD matrix.

File: test/lib/testMayaUsdProxyAccessor.py [hunk 4]
```diff
@@ -0,0 +1,1265 @@
...
+        
+        # Current limitation requires access plugs to be created before we start manipulating and keying
+        translatePlug = pa.GetOrCreateAccessPlug(ufeItemParent, usdAttrName='xformOp:translate')
+        rotatePlug = pa.GetOrCreateAccessPlug(ufeItemParent, usdAttrName='xformOp:rotateXYZ')
+        cmds.currentTime(1) # trigger compute to fill the data
+        
+        cachingScope.waitForCache(self)
+        cachingScope.checkValidFrames(self, self.cache_allFrames)
+        
+        # Manipulate using UFE and key (at frame 1)
```
Comment (ppt-adsk) [inline test/lib/testMayaUsdProxyAccessor.py:923]: using USD

File: lib/mayaUsd/nodes/proxyAccessor.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+        if selectedDagPath == None or selectedPrimPath == None:
+            return None
+
+        sdfPath = Sdf.Path(selectedPrimPath)
+        if not usdAttrName == "":
+            sdfPath = sdfPath.AppendProperty(usdAttrName)
+            
+        plugNameValueAttr = GetAccessPlugName(sdfPath)
+    
+        if not usdAttrName == "":
```
[the remark this answers is not in the document]
Comment (pilarmolinalopez) [reply lib/mayaUsd/nodes/proxyAccessor.py:185]: I agree with @ppt-adsk here

File: lib/mayaUsd/nodes/proxyAccessor.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+# limitations under the License.
+#
+
+"""
+This file contains a set of reference helper functions which can be used to test and discover
+new capabilities provided by proxy accessor on a proxy shape.
+
+The end goal is to leverage UFE to provide integration of mixed data model compute with
+workflows like parenting (USD prim to Maya DAG object or vice versa), keying, or constraining.
+For the time being, this set of helper functions can be used as a reference to understand
```
Comment (pilarmolinalopez) [inline lib/mayaUsd/nodes/proxyAccessor.py:23]: This file was very useful to understand the entire PR and also for testing. Thanks!

File: lib/mayaUsd/nodes/proxyShapeBase.h [hunk 7]
```diff
@@ -211,6 +218,14 @@ class MayaUsdProxyShapeBase : public MPxSurfaceShape,
                 const MEvaluationNode& evaluationNode,
                 PostEvaluationType evalType) override;
 
+#if MAYA_API_VERSION >= 20210000
```
Comment (pilarmolinalopez) [inline lib/mayaUsd/nodes/proxyShapeBase.h:221]: `configCache` and `getCacheSetup` are also available in 2020. why are we allowing only 2021 here?

File: lib/mayaUsd/listeners/stageNoticeListener.h [hunk 8]
```diff
@@ -47,31 +47,38 @@ class UsdMayaStageNoticeListener : public TfWeakBase
...
+        using StageObjectsChangedCallback = std::function<void(const UsdNotice::ObjectsChanged& notice)>;
 
         /// Sets the callback to be invoked when the listener receives a
         /// StageContentsChanged notice.
         MAYAUSD_CORE_PUBLIC
         void SetStageContentsChangedCallback(
-                const StageContentsChangedCallback& callback);
+            const StageContentsChangedCallback& callback);
+
+        MAYAUSD_CORE_PUBLIC
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/listeners/stageNoticeListener.h:62]: Indeed, see 90eccd8f

File: lib/mayaUsd/nodes/proxyAccessor.py [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+
+from pxr import Sdf, Usd, UsdGeom, Vt
+from mayaUsd import lib as mayaUsdLib
+import maya.cmds as cmds
+import maya.OpenMaya as om
+import ufe
+
+def GetUfeSelection():
+    try:
+        return next(iter(ufe.GlobalSelection.get()))
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.py]: Given ProxyAccessor is only enabled in PreviewReleases...then yes, we can switch to the v2 interface.

File: lib/mayaUsd/nodes/proxyAccessor.py [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+
+def GetDagAndPrimFromUfe(ufeObject):
+    if ufeObject == None:
+        return None, None
+        
+    segmentCount = len(ufeObject.path().segments)
+    if segmentCount == 0:
+        return None, None
+        
+    dagPath = str(ufeObject.path().segments[0])[6:]
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.py:47]: The fact that we have to deal with `|world` is something I really don't like. I tried your suggestion but it didn't work as a replacement for the existing code.

[the remark this answers is not in the document]
Comment (ppt-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.py:47]: I agree, I really don't like it either, and that's why I made the suggestion.  I wonder what didn't work; do you have a recent Maya?  "|world" should not be produced by ufe.PathString.string().  Otherwise we can have a look at this when you're back at work on Monday, but rest assured that we will get rid of [6:]...

File: lib/mayaUsd/nodes/proxyAccessor.py [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+    
+    primPath = str(ufeObject.path().segments[1])
+    
+    return dagPath, primPath
+
+def GetSelectedDagAndPrim():    
+    return GetDagAndPrimFromUfe(GetUfeSelection())
+
+def GetAccessPlugName(sdfPath):
+    plugNameValueAttr = 'AccessValue_' + str(sdfPath.__hash__())
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.py:60]: There have been a few iterations around encoding sdf path onto accessor plugs. Many things got improved and we could now get away from using hash...but sdf path will have to be sanitized to remove illegal characters and we need a way to quickly identify accessor plugs (prefix or something else). FWIW. Check what's allowed in attributes names and you will see why we needed to go with nice name to encode the actual path.

It's something we can iterate on in a separate PR...but please share your ideas for the sanitized version we could use as attribute name.

File: lib/mayaUsd/nodes/proxyAccessor.py [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+    
+    return plugNameValueAttr
+
+def IsUfeUsdPath(ufeObject):
+    segmentCount = len(ufeObject.path().segments)
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.py]: Done, see 90eccd8f

File: lib/mayaUsd/nodes/proxyAccessor.py [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+
+    xformAPI = UsdGeom.XformCommonAPI(prim)
+    xformAPI.CreateXformOps(
+                UsdGeom.XformCommonAPI.OpTranslate,
+                UsdGeom.XformCommonAPI.OpRotate,
+                UsdGeom.XformCommonAPI.OpScale)
+
+def CreateAccessPlug(proxyDagPath, sdfPath, sdfValueType):
+    savedSelection = ufe.Selection()
+    savedSelection.replaceWith(ufe.GlobalSelection.get())
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.py]: Leftover from the older implementation which wasn't leveraging ReadUtil.FindOrCreateMayaAttr

File: lib/mayaUsd/nodes/proxyAccessor.py [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+
+    nodeFn = om.MFnDependencyNode(obj)
+    attrObj = nodeFn.attribute(plugNameValueAttr)
+    attrFn = om.MFnTypedAttribute(attrObj)
+    attrFn.setWorldSpace(True)
+        
+    # Set default value
+    cmds.setAttr('{}.{}[0]'.format(proxyDagPath,plugNameValueAttr), [1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0], type='matrix')
+    
+    ufe.GlobalSelection.get().replaceWith(savedSelection)
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.py]: Same as above.

File: lib/mayaUsd/nodes/proxyAccessor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.py:173]: I agree this method is not future proof and totally hacky. I personally would wait with improvements when we will start handling outputs of more things which don't map to attributes (for now we only have world matrix...but I bet there will be more and we will have to figure out how to encode it in the path).

File: lib/mayaUsd/nodes/proxyAccessor.py [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+the concepts necessary to build tooling on top of the proxy accessor.
+"""
+
+from pxr import Sdf, Usd, UsdGeom, Vt
+from mayaUsd import lib as mayaUsdLib
+import maya.cmds as cmds
+import maya.OpenMaya as om
+import ufe
+
+def GetUfeSelection():
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.py]: Now you see how old this code is...written before we had these guidelines. Fixed in 70b88ae44

File: lib/mayaUsd/nodes/proxyAccessor.py [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+    plugNameValueAttr = GetOrCreateAccessPlug(ufeObject,usdAttrName)
+    
+    if plugNameValueAttr == None:
+        print("Error in keying. No attributes to key")
+    else:
+        result = cmds.setKeyframe(proxyDagPath, attribute=plugNameValueAttr)
+    
+def ParentItems(ufeChildren, ufeParent):
+    if not IsUfeUsdPath(ufeParent):
+        print("Last selected item needs to be USD parent")
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.py]: Improved, see 90eccd8f

File: lib/mayaUsd/nodes/proxyAccessor.h [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+
+            accessor->addCallbacks(node.thisMObject());
+
+            return accessor;
+        }
+
+        /*! \brief  Insert extra plug level dependencies for accessor plugs
+            \note   Call from MPxNode::setDependentsDirty()
+         */
+        static MStatus addDependentsDirty(Owner& accessor, const MPlug& plug, MPlugArray& plugArray)
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.h]: Absolutely! See 90eccd8f

File: lib/mayaUsd/nodes/proxyAccessor.h [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+                return accessor->addDependentsDirty(plug, plugArray);
+            else
+                return MStatus::kFailure;
+        }
+
+        /*! \brief  Compute will write input accessor plugs and write converted data to the stage.
+           Once completed, all output accessor plugs will be provided with data. \note   Call from
+           MPxNode::compute()
+         */
+        static MStatus compute(Owner& accessor, const MPlug& plug, MDataBlock& dataBlock)
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.h]: 90eccd8f

File: lib/mayaUsd/nodes/proxyAccessor.h [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+        {
+            if (accessor)
+                return accessor->compute(plug, dataBlock);
+            else
+                return MStatus::kFailure;
+        }
+
+        /*! \brief  Proxy accessor is creating acceleration structure to avoid the discovery of
+         accessor plugs at each compute. This accelleration structure have to be invalidate when
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.h]: 90eccd8f

File: lib/mayaUsd/nodes/proxyAccessor.h [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+         stage changes.
+        */
+        static MStatus
+        stageChanged(Owner& accessor, const MObject& node, const UsdNotice::ObjectsChanged& notice)
```
[the remark this answers is not in the document]
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.h]: 90eccd8f

File: test/lib/testMayaUsdProxyAccessor.py (hunk 2 above)
Comment (kxl-adsk) [reply test/lib/testMayaUsdProxyAccessor.py]: Looks to be working, thx! 90eccd8f

File: test/lib/testMayaUsdProxyAccessor.py (hunk 3 above)
Comment (kxl-adsk) [reply test/lib/testMayaUsdProxyAccessor.py:502]: Yes

File: lib/mayaUsd/nodes/proxyAccessor.py (hunk 6 above)
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.py:23]: Happy to hear it!

File: lib/mayaUsd/nodes/proxyShapeBase.h (hunk 7 above)
Comment (kxl-adsk) [reply lib/mayaUsd/nodes/proxyShapeBase.h:221]: Exactly, ProxyAccessor is only available in PreviewRelease because it depends on a number of fixes done on the Maya side. For this particular case, you are correct that cache interfaces are available in Maya 2020, but nothing will use them and we would still need to add conditional compilation for older versions of Maya. This is why I decided to keep it simple and turn it all off for released versions of Maya.

Comment (pilarmolinalopez) [reply lib/mayaUsd/nodes/proxyShapeBase.h:221]: understood! thanks!

**Comment by kxl-adsk** [comment]
Pushed new changes.

The first change is to properly handle context evaluation. We shouldn't be creating new stages or causing from background context (and threads!) changes to listeners or data consumers.

The second part is about randomly failing caching scope tests. I was hoping that context fixes will make the random failure go away, but they didn't. I got correctness back and warnings removed but we still have a random failure when running the test multiple times. I believe it's a timing problem and fix may have to happen on the Maya side. I won't block this PR since tests are stable and passing in non caching scope (run then 100 times in a row without failure).

**Comment by kxl-adsk** [comment]
@williamkrick can you have a look at b057b59609 (changes I made to proxy shape base to properly support context evaluation).

**Comment by williamkrick** [comment]
> 
> 
> @williamkrick can you have a look at [b057b59](https://github.com/Autodesk/maya-usd/commit/b057b59609d886860270042007a7be328203e3e1) (changes I made to proxy shape base to properly support context evaluation).

It looks okay to me. If we start with the assumption that changing time in USD is similar to cache restore in that it is fast to update the data model then switching the USD stage seems like the correct thing to do.

**Comment by kxl-adsk** [comment]
@williamkrick we actually have no other choice than sharing the stage for context evaluation - consider changes made in foreground (or even simpler case like in-memory root layer) which have to be available when computing in background. Because of this, we have to make all contexts see the same stage.

File: lib/mayaUsd/nodes/proxyAccessor.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ppt-adsk) [reply lib/mayaUsd/nodes/proxyAccessor.py:60]: Let's pick this up separately.
